pkg/postage/listener/listener.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/postage"
+)
+
+const (
+	blockPage = 500 // how many blocks to sync every time
+	tailSize  = 100 // how many blocks to tail from the tip of the chain
+)
+
+var (
+	chainUpdateInterval = 30 * time.Second
+)
+
+type BlockHeightContractFilterer interface {
+	bind.ContractFilterer
+	BlockNumber(context.Context) (uint64, error)
+}
+
+type listener struct {
+	logger          logging.Logger
+	ev              BlockHeightContractFilterer
+	postageStampABI abi.ABI
+	priceOracleABI  abi.ABI
+
+	batchCreatedTopic       common.Hash
+	batchTopupTopic         common.Hash
+	batchDepthIncreaseTopic common.Hash
+	priceUpdateTopic        common.Hash
+
+	postageStampAddress common.Address
+	priceOracleAddress  common.Address
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+func New(
+	logger logging.Logger,
+	ev BlockHeightContractFilterer,
+	postageStampAddress,
+	priceOracleAddress common.Address,
+) postage.Listener {
+	postageStampABI := parseABI(PostageStampABI)
+	priceOracleABI := parseABI(PriceOracleABI)
+	return &listener{
+		logger:                  logger,
+		ev:                      ev,
+		postageStampABI:         postageStampABI,
+		priceOracleABI:          priceOracleABI,
+		batchCreatedTopic:       postageStampABI.Events["BatchCreated"].ID,
+		batchTopupTopic:         postageStampABI.Events["BatchTopUp"].ID,
+		batchDepthIncreaseTopic: postageStampABI.Events["BatchDepthIncrease"].ID,
+		priceUpdateTopic:        priceOracleABI.Events["PriceUpdate"].ID,
+
+		postageStampAddress: postageStampAddress,
+		priceOracleAddress:  priceOracleAddress,
+	}
+}
+
+func (l *listener) Listen(from uint64, updater postage.EventUpdater) error {
+	l.wg.Add(1)
+
+	go func() {
+		defer l.wg.Done()
+		err := l.sync(from, updater)
+		if err != nil {
+			l.logger.Errorf("event listener sync: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (l *listener) filterQuery(from, to *big.Int) ethereum.FilterQuery {
+	return ethereum.FilterQuery{
+		FromBlock: from,
+		ToBlock:   to,
+		Addresses: []common.Address{
+			l.postageStampAddress,
+			l.priceOracleAddress,
+		},
+		Topics: [][]common.Hash{
+			{
+				l.batchCreatedTopic,
+				l.batchTopupTopic,
+				l.batchDepthIncreaseTopic,
+				l.priceUpdateTopic,
+			},
+		},
+	}
+}
+
+func (l *listener) parseEvent(a *abi.ABI, eventName string, c interface{}, e types.Log) error {
+	err := a.Unpack(c, eventName, e.Data)
+	if err != nil {
+		return err
+	}
+
+	var indexed abi.Arguments
+	for _, arg := range a.Events[eventName].Inputs {
+		if arg.Indexed {
+			indexed = append(indexed, arg)
+		}
+	}
+	return abi.ParseTopics(c, indexed, e.Topics[1:])
+}
+
+func (l *listener) processEvent(e types.Log, updater postage.EventUpdater) error {
+	eventSig := e.Topics[0]
+	switch eventSig {
```

Comment:
`switch e.Topics[0] {`

Reply:
done, but i do think the previous way would make it more clear to the reader what this switch is actually doing.

Reply:
ok fair enough :+1: